### PR TITLE
Adapt transform to C++20

### PR DESCRIPTION
**Pull request by gonidelis**
This adapts `transform` to C++20 (all overloads according to [the Standard](http://eel.is/c++draft/alg.transform)) and adds unit tests. It deprecates the `hpx::parallel::transform` overloads.

flyby: it adds `get_in_in_out_result`.

### CODE REVIEW

File: libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt
Comment (msimberg) [inline libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt:89]: Why is this commented out?

**Review by msimberg — COMMENTED** [review]
Thanks @gonidelis for updating this!

File: libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt:89]: This test checks an overload that does not exist in the updated C++20 overloads. There is some discussion need to be made as far as tests are concerned. I added `transform_range2` and I think there are one or two tests that need to be added in order for all the overloads to be checked. Please tell me if you agree in order for me to completeley delete `transform_range_binary`.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform.hpp
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform.hpp:722]: Actually I have already manually formated these, so please tell me if there is any problem with the identation and I will fix the `&&`s asap. Thanks :)

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform.hpp
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform.hpp:13]: I will check it and update you soon. Thank you :)

Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform.hpp:13]: @msimberg  Seems like it is redundant after all. It will be removed in the next commit.

File: libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt
Comment (msimberg) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt]: Sorry for not responding here earlier @gonidelis. I think it'd make sense to leave the test for the overloads that are not part of C++20. This means they'd still be using `hpx::parallel::transform`, `hpx::transform` or `hpx::ranges::transform`. This will lead to deprecation warnings so you'd have to add either `#define HPX_HAVE_DEPRECATION_WARNINGS_V1_6 0` to the test, or `target_compile_definitions(transform_range_binary_test HPX_HAVE_DEPRECATION_WARNINGS_V1_6=0)` to the CMake file. This will of course be deleted once we delete the deprecated overloads but it's unnecessary to delete tests right now that we already have.

**Comment by gonidelis** [comment]
@hkaiser  Thank you very much for providing this!

File: libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt
Comment (gonidelis) [reply libs/parallelism/algorithms/tests/unit/container_algorithms/CMakeLists.txt]: @msimberg  sorry for responding so late to this, but since we are going to leave just this test as is and since `parallel:transform` invokes `transform_` which returns an `in_out_result` there is some inconsistency here. I could change the result type of `parallel::transform` but it wouldn't be deprecated then. What's the trick?

**Comment by hkaiser** [comment]
@gonidelis apparently, tests/performance/local/CMakeFiles/stream_test.dir/stream.cpp relies on a now obsolete API. Would you mind updating this to the appropriate new API?

File: libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/transform.hpp
Comment (hkaiser) [inline libs/full/segmented_algorithms/include/hpx/parallel/segmented_algorithms/transform.hpp:14]: Are the `tagged_xxx.hpp` headers still needed?

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform.hpp
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform.hpp:400]: Are the `tagged_xxx.hpp` headers still needed?

**Review by hkaiser — DISMISSED** [review]
Otherwise, LGTM, thanks!

**Comment by msimberg** [comment]
@gonidelis I think this is the last remaining deprecated overload that should be updated: https://github.com/gonidelis/hpx/blob/09bff92fb2be501e89dd6437e524091d150d2bc0/libs/full/compute_cuda/tests/unit/transform_compute.cu#L48-L49.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform.hpp
Comment (gonidelis) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform.hpp:400]: No they are not. Please don't merge there is still some minor housekeeping left done. It will be ready within the day. Thanks :)

**Comment by gonidelis** [comment]
@msimberg  Ok I will fix that. Is CI back again? I will do a `grep` just to be safe in any other case.

**Comment by msimberg** [comment]
> @msimberg Ok I will fix that. Is CI up again? I will do a `grep` just to be safe in any other case.

The rostam jenkins CI is still down, but all the rest are up as usual. There are some additional configurations on rostam, but they're unlikely to uncover any additional issues on this PR.

**Comment by gonidelis** [comment]
I am pretty sure that I fixed any remaining depracations and unnecessary `tagged_xxx` uses. The PR should be ready to merge now if CI is ok with it.

**Comment by msimberg** [comment]
Looks like CI is (sort of) happy now, thanks @gonidelis! Let's merge this and deal with any remaining problems once rostam is back up (hopefully none).

**Pull request merged**
